### DISCUSSION TITLE
Disappearing spaces in transcription of speech to text one way but reappearing the other way

### DISCUSSION THREAD

**Discussion by catskillsresearch** [NVIDIA-NeMo/NeMo#2898] [Q&A] [answered]
I have a BPE tokenizer with the ##'s following the pattern of the [ASR_with_subword_tokenization tutorial ](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/ASR_with_Subword_Tokenization.ipynb) notebook.   If I transcribe a text using this snippet from [examples/asr/speech_to_text_infer.py](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/speech_to_text_infer.py):

```
    hypotheses = []
    references = []
    for test_batch in asr_model.test_dataloader():
        if can_gpu:
            test_batch = [x.cuda() for x in test_batch]
        with autocast():
            log_probs, encoded_len, greedy_predictions = asr_model(
                input_signal=test_batch[0], input_signal_length=test_batch[1]
            )
        hypotheses += wer.ctc_decoder_predictions_tensor(greedy_predictions)
        for batch_ind in range(greedy_predictions.shape[0]):
            seq_len = test_batch[3][batch_ind].cpu().detach().numpy()
            seq_ids = test_batch[2][batch_ind].cpu().detach().numpy()
            reference = ''.join([labels_map[c] for c in seq_ids[0:seq_len]])
            references.append(reference)
        del test_batch
```
then my transcription has no space characters in both the references and the hypotheses, even though what the dataloader is loading has spaces in the 'text' fields in the manifest..   On the other hand, for the exact same model, tokenizer and inputs, if I instead transcribe like this:

`hypotheses += asr_model.transcribe(paths2audio_files=audio_fns,batch_size=512)`

then my transcriptions have the spaces inserted in the correct place.

What's going on here?  Is there a way to correct the first snippet so that it doesn't lose spaces?  I'm much happier in some ways to do it as a one-liner, there is a lot less setup involved, but I'm mystified about where my spaces are going in the first snippet.

**Comment by catskillsresearch** [comment, marked as answer]
I think I figured it out: In subword encoding, by it's name, only subwords are encoded.  There is an implied space between each subword except when two adjacent subwords are matched with ##.  Then the two subwords can be joined and there is no space.

**Reply by titu1994** [reply]
Sorry for delay. As you've guessed correctly, different tokenization schemes have different ways to represent space. Sentencepiece on particular uses _ prior to subword to represent a space prior to a subword. 

You'll note that vocab.txt generated by Nemo is just a pedantic mapping from the Sentencepiece vocab to a "HF-like" vocab used for quick debugging. The actual tokenizer used is the tokenizer.model
